### DISCUSSION TITLE
Instrumentation extension service

### DISCUSSION THREAD

**Discussion by BFergerson** [apache/skywalking#9733] [Ideas]
I was looking to see what the easiest way to add stack traces to exit spans was and stumbled upon the [extension](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension) system offered in OpenTelemetry. I'm curious if this is a feature SkyWalking would consider adding. I know my use case is unique, so here is another that may be more universal.

In the manual instrumentation documentation, the following example is given:
```java
@Trace
@Tag(key = "tag1", value = "arg[0]")
@Tag(key = "tag2", value = "arg[1]")
@Tag(key = "username", value = "returnedObj.username")
@Tag(key = "age", value = "returnedObj.age")
public User methodYouWantToTrace(String param1, String param2) {
    // ActiveSpan.setOperationName("Customize your own operation name, if this is an entry span, this would be an endpoint name");
    // ...
}
```

I like the idea of adding the `username` and `age` to `methodYouWantToTrace`, but what if I wanted to add those tags to any EntrySpan starting with the URI `/settings/profile/*` so that they are automatically inserted whenever users make changes to their profile?

In my last discussion in which I brought up the possibility of creating a gauge for usernames, that was determined to be business monitoring. Does this also fall under that umbrella? If it does, would SkyWalking consider supporting it? If not, is there an alternative?

**Comment by wu-sheng** [comment]
This use case seems closely to APM case. I want to hear your suppose first.
What is your suggested way to identify this when profiling?

**Reply by BFergerson** [reply]
What do you mean by my suppose? I'm also not sure what profiling has to do with this. This would be an extension of the tracing functionality. `ContextManagerExtendService` looks pretty close to already providing a way to extend tracing. You would just need to override `createTraceContext` and provide a custom `AbstractTracerContext`.

I'm suggesting offering an easier alternative that would have listeners for each span type. Users could then do something like:
```java
@InstrumentationExtension
class ProfileAttrInjector extends UserInstrumentationInjector {
    public void onEntrySpan(EntrySpan span) {
        if (span.getTag("url") != null && span.getTag("url").startsWith("/settings/profile/")) {
            span.tag("username", *username*);
            span.tag("age", *age*);
        }
    }
}
```

**Reply by wu-sheng** [reply]
I think this could be easier.

1. Use `Correlation` APIs to inject this into the trace context, https://skywalking.apache.org/docs/skywalking-java/next/en/setup/service-agent/java-agent/application-toolkit-trace/
2. Open `correlation.auto_tag_keys` property(Java agent only) to make Correlation context in tags automatically.

Of course, your ways works too.

**Reply by BFergerson** [reply]
That could work too. One thought I did have after that message was providing `(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret)` types. That way if there are values being passed to a method, they could use them to add tags to the current span.

That way this would be just as powerful as the `@Tag(key = "username", value = "returnedObj.username")` style.

Without the EnhancedInstance and arguments, they would need to ensure the have some context to pull data from and would probably end up just going with the `@Tag` style anyway. Thoughts?

**Reply by wu-sheng** [reply]
I am not sure about `providing (EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret) types`. How are we going to provide this?

**Reply by BFergerson** [reply]
I imagine SkyWalking scans the classpath for `@InstrumentationExtension`, creates instances of those classes, and calls them during `InstMethodsInter.intercept`

From my understanding, the active span would be open after `beforeMethod` and closed after `afterMethod`. So this code would go somewhere in between those.

**Reply by wu-sheng** [reply]
This seems a very expensive way. I am wondering what could we do in a general interceptor?

> I imagine SkyWalking scans the classpath for @InstrumentationExtension

By the way, this may not work. The user's classpath may not be a good place for an interceptor-level calling, due to classpath isolation.
Something like this has to be a thing similar with today's plugin codes, from tech perspective.

**Reply by BFergerson** [reply]
I think the idea of a general interceptor works if it can be done inside user code as opposed to creating a new plugin and adding it to `/plugins`. SkyWalking plugin system is powerful but a bit difficult to get started with. If someone wants to add instrumentation for a library that SW doesn't support they have to create a new plugin, use `apm-customize-enhance-plugin`, or rebuild the library with `@Trace` and other code to the library.

This is probably my view but I would like to see SW be more programmable (i.e. observability-as-code). I would like to simply create a new class, add some annotations, and deploy my code and it's like I've added a new plugin. Maybe that's too much to ask though. I'm not sure what the limitations are.

As for the classpath issue. If SW can find `@Trace`, why couldn't it find a `@InstrumentationExtension`?

If that's not possible, could SW offer something like:
 ```java
SkyWalkingAgent.addUserPlugin(ProfileAttrInjector.class);
```

**Reply by wu-sheng** [reply]
There is no block to finding this class, the challenge is running that class. Class definition is scoped inside the classloader, you can't initial a class inside a classloader(a), but running in another classloader(b), unless classloader(b)  is a child of classloader(a).

**Reply by BFergerson** [reply]
That makes sense. I have a closer look at how OpenTelemetry extensions work and let you know if I come up with a better idea.

**Reply by wu-sheng** [reply]
This mechanism would be easy to work on Spring boot or other single Classloader env. But would be tricky if you work in Tomcat or similar mechanism.
